Crash Report 3.6: Fatal Exception: java.lang.RuntimeException

java.lang.RuntimeException: Parcelable encountered IOException writing serializable object (name = com.android.volley.ServerError)
       at android.os.Parcel.writeSerializable(Parcel.java:1316)
       at android.os.Parcel.writeValue(Parcel.java:1264)
       at android.os.Parcel.writeArray(Parcel.java:672)
       at android.os.Parcel.writeValue(Parcel.java:1251)
       at android.os.Parcel.writeArrayMapInternal(Parcel.java:618)
       at android.os.Bundle.writeToParcel(Bundle.java:1692)
       at android.os.Parcel.writeBundle(Parcel.java:636)
       at android.app.FragmentState.writeToParcel(Fragment.java:133)
       at android.os.Parcel.writeTypedArray(Parcel.java:1133)
       at android.app.FragmentManagerState.writeToParcel(FragmentManager.java:373)
       at android.os.Parcel.writeParcelable(Parcel.java:1285)
       at android.os.Parcel.writeValue(Parcel.java:1204)
       at android.os.Parcel.writeArrayMapInternal(Parcel.java:618)
       at android.os.Bundle.writeToParcel(Bundle.java:1692)
       at android.os.Parcel.writeBundle(Parcel.java:636)
       at android.app.ActivityManagerProxy.activityStopped(ActivityManagerNative.java:2755)
       at android.app.ActivityThread$StopInfo.run(ActivityThread.java:3396)
       at android.os.Handler.handleCallback(Handler.java:733)
       at android.os.Handler.dispatchMessage(Handler.java:95)
       at android.os.Looper.loop(Looper.java:136)
       at android.app.ActivityThread.main(ActivityThread.java:5579)
       at java.lang.reflect.Method.invokeNative(Method.java)
       at java.lang.reflect.Method.invoke(Method.java:515)
       at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:1268)
       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1084)
       at dalvik.system.NativeStart.main(NativeStart.java)
Caused by: java.io.NotSerializableException: com.android.volley.NetworkResponse
       at java.io.ObjectOutputStream.writeNewObject(ObjectOutputStream.java:1364)
       at java.io.ObjectOutputStream.writeObjectInternal(ObjectOutputStream.java:1671)
       at java.io.ObjectOutputStream.writeObject(ObjectOutputStream.java:1517)
       at java.io.ObjectOutputStream.writeObject(ObjectOutputStream.java:1481)
       at java.io.ObjectOutputStream.writeFieldValues(ObjectOutputStream.java:979)
       at java.io.ObjectOutputStream.defaultWriteObject(ObjectOutputStream.java:368)
       at java.io.ObjectOutputStream.writeHierarchy(ObjectOutputStream.java:1074)
       at java.io.ObjectOutputStream.writeNewObject(ObjectOutputStream.java:1404)
       at java.io.ObjectOutputStream.writeObjectInternal(ObjectOutputStream.java:1671)
       at java.io.ObjectOutputStream.writeObject(ObjectOutputStream.java:1517)
       at java.io.ObjectOutputStream.writeObject(ObjectOutputStream.java:1481)
       at android.os.Parcel.writeSerializable(Parcel.java:1311)
       at android.os.Parcel.writeValue(Parcel.java:1264)
       at android.os.Parcel.writeArray(Parcel.java:672)
       at android.os.Parcel.writeValue(Parcel.java:1251)
       at android.os.Parcel.writeArrayMapInternal(Parcel.java:618)
       at android.os.Bundle.writeToParcel(Bundle.java:1692)
       at android.os.Parcel.writeBundle(Parcel.java:636)
       at android.app.FragmentState.writeToParcel(Fragment.java:133)
       at android.os.Parcel.writeTypedArray(Parcel.java:1133)
       at android.app.FragmentManagerState.writeToParcel(FragmentManager.java:373)
       at android.os.Parcel.writeParcelable(Parcel.java:1285)
       at android.os.Parcel.writeValue(Parcel.java:1204)
       at android.os.Parcel.writeArrayMapInternal(Parcel.java:618)
       at android.os.Bundle.writeToParcel(Bundle.java:1692)
       at android.os.Parcel.writeBundle(Parcel.java:636)
       at android.app.ActivityManagerProxy.activityStopped(ActivityManagerNative.java:2755)
       at android.app.ActivityThread$StopInfo.run(ActivityThread.java:3396)
       at android.os.Handler.handleCallback(Handler.java:733)
       at android.os.Handler.dispatchMessage(Handler.java:95)
       at android.os.Looper.loop(Looper.java:136)
       at android.app.ActivityThread.main(ActivityThread.java:5579)
       at java.lang.reflect.Method.invokeNative(Method.java)
       at java.lang.reflect.Method.invoke(Method.java:515)
       at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:1268)
       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1084)
       at dalvik.system.NativeStart.main(NativeStart.java)

I bet this is something that happens in Stats :/ 
(Caused by: java.io.NotSerializableException: com.android.volley.NetworkResponse)

java.lang.RuntimeException: Parcelable encountered IOException writing serializable object (name = [Ljava.io.Serializable;)
       at android.os.Parcel.writeSerializable(Parcel.java:1388)
       at android.os.Parcel.writeValue(Parcel.java:1335)
       at android.os.Parcel.writeArrayMapInternal(Parcel.java:638)
       at android.os.BaseBundle.writeToParcelInner(BaseBundle.java:1313)
       at android.os.Bundle.writeToParcel(Bundle.java:1096)
       at android.os.Parcel.writeBundle(Parcel.java:663)
       at android.app.FragmentState.writeToParcel(Fragment.java:139)
       at android.os.Parcel.writeTypedArray(Parcel.java:1191)
       at android.app.FragmentManagerState.writeToParcel(FragmentManager.java:376)
       at android.os.Parcel.writeParcelable(Parcel.java:1357)
       at android.os.Parcel.writeValue(Parcel.java:1262)
       at android.os.Parcel.writeArrayMapInternal(Parcel.java:638)
       at android.os.BaseBundle.writeToParcelInner(BaseBundle.java:1313)
       at android.os.Bundle.writeToParcel(Bundle.java:1096)
       at android.os.Parcel.writeBundle(Parcel.java:663)
       at android.app.ActivityManagerProxy.activityStopped(ActivityManagerNative.java:2884)
       at android.app.ActivityThread$StopInfo.run(ActivityThread.java:3261)
       at android.os.Handler.handleCallback(Handler.java:739)
       at android.os.Handler.dispatchMessage(Handler.java:95)
       at android.os.Looper.loop(Looper.java:135)
       at android.app.ActivityThread.main(ActivityThread.java:5221)
       at java.lang.reflect.Method.invoke(Method.java)
       at java.lang.reflect.Method.invoke(Method.java:372)
       at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:899)
       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:694)
Caused by: java.io.NotSerializableException: com.android.volley.NetworkResponse
       at java.io.ObjectOutputStream.writeNewObject(ObjectOutputStream.java:1344)
       at java.io.ObjectOutputStream.writeObjectInternal(ObjectOutputStream.java:1651)
       at java.io.ObjectOutputStream.writeObject(ObjectOutputStream.java:1497)
       at java.io.ObjectOutputStream.writeObject(ObjectOutputStream.java:1461)
       at java.io.ObjectOutputStream.writeFieldValues(ObjectOutputStream.java:959)
       at java.io.ObjectOutputStream.defaultWriteObject(ObjectOutputStream.java:360)
       at java.io.ObjectOutputStream.writeHierarchy(ObjectOutputStream.java:1054)
       at java.io.ObjectOutputStream.writeNewObject(ObjectOutputStream.java:1384)
       at java.io.ObjectOutputStream.writeObjectInternal(ObjectOutputStream.java:1651)
       at java.io.ObjectOutputStream.writeObject(ObjectOutputStream.java:1497)
       at java.io.ObjectOutputStream.writeObject(ObjectOutputStream.java:1461)
       at java.io.ObjectOutputStream.writeNewArray(ObjectOutputStream.java:1185)
       at java.io.ObjectOutputStream.writeObjectInternal(ObjectOutputStream.java:1642)
       at java.io.ObjectOutputStream.writeObject(ObjectOutputStream.java:1497)
       at java.io.ObjectOutputStream.writeObject(ObjectOutputStream.java:1461)
       at android.os.Parcel.writeSerializable(Parcel.java:1383)
       at android.os.Parcel.writeValue(Parcel.java:1335)
       at android.os.Parcel.writeArrayMapInternal(Parcel.java:638)
       at android.os.BaseBundle.writeToParcelInner(BaseBundle.java:1313)
       at android.os.Bundle.writeToParcel(Bundle.java:1096)
       at android.os.Parcel.writeBundle(Parcel.java:663)
       at android.app.FragmentState.writeToParcel(Fragment.java:139)
       at android.os.Parcel.writeTypedArray(Parcel.java:1191)
       at android.app.FragmentManagerState.writeToParcel(FragmentManager.java:376)
       at android.os.Parcel.writeParcelable(Parcel.java:1357)
       at android.os.Parcel.writeValue(Parcel.java:1262)
       at android.os.Parcel.writeArrayMapInternal(Parcel.java:638)
       at android.os.BaseBundle.writeToParcelInner(BaseBundle.java:1313)
       at android.os.Bundle.writeToParcel(Bundle.java:1096)
       at android.os.Parcel.writeBundle(Parcel.java:663)
       at android.app.ActivityManagerProxy.activityStopped(ActivityManagerNative.java:2884)
       at android.app.ActivityThread$StopInfo.run(ActivityThread.java:3261)
       at android.os.Handler.handleCallback(Handler.java:739)
       at android.os.Handler.dispatchMessage(Handler.java:95)
       at android.os.Looper.loop(Looper.java:135)
       at android.app.ActivityThread.main(ActivityThread.java:5221)
       at java.lang.reflect.Method.invoke(Method.java)
       at java.lang.reflect.Method.invoke(Method.java:372)
       at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:899)
       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:694)

--- a/WordPress/src/main/java/org/wordpress/android/ui/stats/StatsAbstractListFragment.java
+++ b/WordPress/src/main/java/org/wordpress/android/ui/stats/StatsAbstractListFragment.java
@@ -131,6 +131,16 @@ public abstract class StatsAbstractListFragment extends StatsAbstractFragment {
     @Override
     public void onSaveInstanceState(Bundle outState) {
         //AppLog.d(AppLog.T.STATS, this.getTag() + " > saving instance state");
+
+        // Do not serialize VolleyError. FIX for https://github.com/wordpress-mobile/WordPress-Android/issues/2228
+        if (mDatamodels != null) {
+            for (int i=0; i < mDatamodels.length; i++) {
+                if (mDatamodels[i] instanceof VolleyError) {
+                    mDatamodels[i] = null;
+                }
+            }
+        }
+
         outState.putSerializable(ARG_REST_RESPONSE, mDatamodels);
         super.onSaveInstanceState(outState);
     }

--- a/WordPress/src/main/java/org/wordpress/android/ui/stats/StatsAbstractListFragment.java
+++ b/WordPress/src/main/java/org/wordpress/android/ui/stats/StatsAbstractListFragment.java
@@ -23,6 +23,7 @@ import com.android.volley.NoConnectionError;
 import com.android.volley.VolleyError;
 
 import org.wordpress.android.R;
+import org.wordpress.android.ui.stats.exceptions.StatsError;
 import org.wordpress.android.ui.stats.service.StatsService;
 import org.wordpress.android.util.AppLog;
 import org.wordpress.android.util.DisplayUtils;
@@ -132,11 +133,14 @@ public abstract class StatsAbstractListFragment extends StatsAbstractFragment {
     public void onSaveInstanceState(Bundle outState) {
         //AppLog.d(AppLog.T.STATS, this.getTag() + " > saving instance state");
 
-        // Do not serialize VolleyError. FIX for https://github.com/wordpress-mobile/WordPress-Android/issues/2228
+        // Do not serialize VolleyError, but rewrite in a simple stats Exception.
+        // VolleyErrors should be serializable, but for some reason they are not.
+        // FIX for https://github.com/wordpress-mobile/WordPress-Android/issues/2228
         if (mDatamodels != null) {
             for (int i=0; i < mDatamodels.length; i++) {
-                if (mDatamodels[i] instanceof VolleyError) {
-                    mDatamodels[i] = null;
+                if (mDatamodels[i] != null && mDatamodels[i] instanceof VolleyError) {
+                    VolleyError currentVolleyError = (VolleyError) mDatamodels[i];
+                    mDatamodels[i] = StatsUtils.rewriteVolleyError(currentVolleyError, getString(R.string.error_refresh_stats));
                 }
             }
         }
@@ -268,7 +272,7 @@ public abstract class StatsAbstractListFragment extends StatsAbstractFragment {
 
     protected boolean isErrorResponse(int index) {
         return mDatamodels != null && mDatamodels[index] != null
-                && mDatamodels[index] instanceof VolleyError;
+                && (mDatamodels[index] instanceof VolleyError || mDatamodels[index] instanceof StatsError);
     }
 
     protected boolean isSingleView() {

--- a/WordPress/src/main/java/org/wordpress/android/ui/stats/StatsUtils.java
+++ b/WordPress/src/main/java/org/wordpress/android/ui/stats/StatsUtils.java
@@ -4,12 +4,15 @@ import android.annotation.SuppressLint;
 import android.content.SharedPreferences;
 import android.preference.PreferenceManager;
 
+import com.android.volley.VolleyError;
+
 import org.json.JSONException;
 import org.json.JSONObject;
 import org.wordpress.android.R;
 import org.wordpress.android.WordPress;
 import org.wordpress.android.WordPressDB;
 import org.wordpress.android.models.Blog;
+import org.wordpress.android.ui.stats.exceptions.StatsError;
 import org.wordpress.android.ui.stats.models.AuthorsModel;
 import org.wordpress.android.ui.stats.models.ClicksModel;
 import org.wordpress.android.ui.stats.models.CommentFollowersModel;
@@ -335,6 +338,24 @@ public class StatsUtils {
                 break;
         }
         return model;
+    }
+
+    /*
+     * This function rewrites a VolleyError into a simple Stats Error by getting the error message.
+     * This is a FIX for https://github.com/wordpress-mobile/WordPress-Android/issues/2228 where
+     * VolleyErrors cannot be serializable.
+     */
+    public static StatsError rewriteVolleyError(VolleyError volleyError, String defaultErrorString) {
+        if (volleyError != null && volleyError.getMessage() != null) {
+            return new StatsError(volleyError.getMessage());
+        }
+
+        if (defaultErrorString != null) {
+            return new StatsError(defaultErrorString);
+        }
+
+        // Error string should be localized here, but don't want to pass a context
+        return new StatsError("Stats couldn't be refreshed at this time");
     }
 
 }

--- a/WordPress/src/main/java/org/wordpress/android/ui/stats/StatsVisitorsAndViewsFragment.java
+++ b/WordPress/src/main/java/org/wordpress/android/ui/stats/StatsVisitorsAndViewsFragment.java
@@ -22,6 +22,7 @@ import com.jjoe64.graphview.GraphViewSeries;
 
 import org.wordpress.android.R;
 import org.wordpress.android.WordPress;
+import org.wordpress.android.ui.stats.exceptions.StatsError;
 import org.wordpress.android.ui.stats.models.VisitModel;
 import org.wordpress.android.ui.stats.models.VisitsModel;
 import org.wordpress.android.ui.stats.service.StatsService;
@@ -166,6 +167,11 @@ public class StatsVisitorsAndViewsFragment extends StatsAbstractFragment
     public void onSaveInstanceState(Bundle outState) {
         //AppLog.d(T.STATS, "StatsVisitorsAndViewsFragment > saving instance state");
 
+        // FIX for https://github.com/wordpress-mobile/WordPress-Android/issues/2228
+        if (mVisitsData != null && mVisitsData instanceof VolleyError) {
+            VolleyError currentVolleyError = (VolleyError) mVisitsData;
+            mVisitsData = StatsUtils.rewriteVolleyError(currentVolleyError, getString(R.string.error_refresh_stats));
+        }
         outState.putSerializable(ARG_REST_RESPONSE, mVisitsData);
         outState.putInt(ARG_SELECTED_GRAPH_BAR, mSelectedBarGraphBarIndex);
         outState.putInt(ARG_SELECTED_OVERVIEW_ITEM, mSelectedOverviewItemIndex);
@@ -214,7 +220,7 @@ public class StatsVisitorsAndViewsFragment extends StatsAbstractFragment
             return;
         }
 
-        if (mVisitsData instanceof VolleyError) {
+        if (mVisitsData instanceof VolleyError || mVisitsData instanceof StatsError) {
             setupNoResultsUI(false);
             return;
         }

--- a/WordPress/src/main/java/org/wordpress/android/ui/stats/exceptions/StatsError.java
+++ b/WordPress/src/main/java/org/wordpress/android/ui/stats/exceptions/StatsError.java
@@ -1,0 +1,9 @@
+package org.wordpress.android.ui.stats.exceptions;
+
+import java.io.Serializable;
+
+public class StatsError extends Exception implements Serializable {
+    public StatsError(String errorMessage) {
+        super(errorMessage);
+    }
+}

